### PR TITLE
Fixed #29251 -- Added CharField converter to the MySQL backend.

### PR DESCRIPTION
**Pull request by felixxm**
https://code.djangoproject.com/ticket/29251

I'm not convinced that this is the best solution, but I haven't found a better way to solve this issue.

### CODE REVIEW

File: django/db/backends/mysql/operations.py [hunk 1]
```diff
@@ -232,10 +232,17 @@ def get_db_converters(self, expression):
...
+        elif internal_type == 'CharField':
+            converters.append(self.convert_charfield_value)
         elif internal_type == 'UUIDField':
             converters.append(self.convert_uuidfield_value)
         return converters
 
+    def convert_charfield_value(self, value, expression, connection):
+        if isinstance(value, bytes):
+            return force_text(value)
+        return value
```
Comment (pope1ni) [inline django/db/backends/mysql/operations.py]: I guess this does follow other existing methods, but:
```python
return force_text(value) if isinstance(value, bytes) else value
```

Comment (felixxm) [reply django/db/backends/mysql/operations.py]: Thanks. I updated this line.

File: tests/db_functions/tests.py [hunk 2]
```diff
@@ -671,3 +671,8 @@ def test_function_as_filter(self):
             Author.objects.exclude(alias=Upper(V('smithj'))),
             ['Rhonda'], lambda x: x.name
         )
+
+    def test_funct_string_combination(self):
+        Author.objects.create(name='Rhonda', alias='john_smith')
+        authors = Author.objects.annotate(filled=LPad('name', Length('alias'), output_field=CharField()))
+        self.assertQuerysetEqual(authors, ['    Rhonda'], lambda a: a.filled, ordered=False)
```
Comment (timgraham) [inline tests/db_functions/tests.py]: I wonder why `ordered=False` with a result of length 1.

File: tests/db_functions/tests.py [hunk 3]
```diff
@@ -671,3 +671,8 @@ def test_function_as_filter(self):
             Author.objects.exclude(alias=Upper(V('smithj'))),
             ['Rhonda'], lambda x: x.name
         )
+
+    def test_funct_string_combination(self):
```
Comment (timgraham) [inline tests/db_functions/tests.py]: funct -> func -- I think? or "function"

Comment (felixxm) [reply tests/db_functions/tests.py]: Typo :man_facepalming: Thanks.

**Comment by felixxm** [comment]
@timgraham Thanks for the review. You're right `decode()` is enough.

**Comment by pope1ni** [comment]
I note that `convert_textfield_value()` below is using `force_text()`.
Perhaps that can be resolved in a follow up commit if `decode()` is enough there too?

**Comment by timgraham** [comment]
Yes, there's [a ticket](https://code.djangoproject.com/ticket/27795) to audit `force_text()` usage.

**Comment by timgraham** [comment]
Then non-ASCII character case doesn't work on Oracle (leading spaces are missing).

**Comment by felixxm** [comment]
`LPad`/`Rpad` doesn't work properly with multibyte characters sets on Oracle. I found the third issue in a row :smile:

**Comment by felixxm** [comment]
@timgraham I checked and IMO there doesn't exist a good workaround for `LPad`/`RPad` with multibyte characters on Oracle. None of them works in all cases. According Oracle [documentation](https://docs.oracle.com/database/121/SQLRF/functions173.htm#SQLRF06103):

_... However, in some multibyte character sets, the display length of a character string can differ from the number of characters in the string._

I think it's fine to add a note in documentation about this restriction.

**Comment by timgraham** [comment]
Sure, if you think is deserves a mention. We can't document everything about each database. :-)

**Comment by felixxm** [comment]
You're right, it's impossible to describe all Oracle's caveats :wink:  I've left a note in the ticket [28643](https://code.djangoproject.com/ticket/28643), this should be enough. I updated tests, it works on Oracle.

**Comment by claudep** [comment]
Would it be worth benchmarking the speed overhead of this addition?

**Comment by felixxm** [comment]
@claudep I wrote simple benchmark with `djangobench`:
```log
Running benchmarks: mysql_char_converter
Control: Django 2.1.dev20180326173423 (in git branch master)
Experiment: Django 2.1.dev20180327171917 (in git branch issue-29251)

Running 'mysql_char_converter' benchmark ...
Min: 0.038486 -> 0.046070: 1.1970x slower
Avg: 0.043529 -> 0.051424: 1.1814x slower
Significant (t=-10.273530)
Stddev: 0.00555 -> 0.00532: 1.0428x smaller (N = 100)

Control: Django 2.1.dev20180326173423 (in git branch master)
Experiment: Django 2.1.dev20180327171917 (in git branch issue-29251)

Running 'mysql_char_converter' benchmark ...
Min: 0.034216 -> 0.042022: 1.2281x slower
Avg: 0.038185 -> 0.045879: 1.2015x slower
Significant (t=-14.931398)
Stddev: 0.00528 -> 0.00502: 1.0528x smaller (N = 200)
```

**Comment by claudep** [comment]
Oh thanks! So does that mean we are ~20% slower with the patch? (sorry, I'm not so great at stats)

**Comment by felixxm** [comment]
I think so, but even if `convert_charfield_value()` doesn't check or change anything I still get about ~11% , i.e. with
```python
def convert_charfield_value(self, value, expression, connection):
    return value
```
the result is
```log
Running benchmarks: mysql_char_converter
Control: Django 2.1.dev20180326173423 (in git branch master)
Experiment: Django 2.1.dev20180328202753 (in git branch issue-29251)

Running 'mysql_char_converter' benchmark ...
Min: 0.075977 -> 0.085019: 1.1190x slower
Avg: 0.086817 -> 0.096318: 1.1094x slower
Significant (t=-8.154523)
Stddev: 0.00763 -> 0.00881: 1.1550x larger (N = 100)

Control: Django 2.1.dev20180326173423 (in git branch master)
Experiment: Django 2.1.dev20180328202753 (in git branch issue-29251)

Running 'mysql_char_converter' benchmark ...
Min: 0.080404 -> 0.091563: 1.1388x slower
Avg: 0.091934 -> 0.102306: 1.1128x slower
Significant (t=-12.184438)
Stddev: 0.00919 -> 0.00777: 1.1821x smaller (N = 200)
```

**Comment by claudep** [comment]
That's probably the overhead cost of adding converters (function calls et al). Then I think we should find another way of solving this issue. Slowing the whole stack just to fix this edge case is not very sound IMHO.

**Comment by felixxm** [comment]
I've prepared alternative PR #9835 with conversion only in affected database functions.

**Comment by felixxm** [comment]
Closing due to the alternative approach in #9835.

**Pull request closed without merging**